network/src/testutils/fake_socket.rs

```diff
@@ -75,8 +75,12 @@ impl<'a> AsyncRead for ReadOnlyTestSocket<'a> {
     fn poll_read(
         mut self: Pin<&mut Self>,
         mut _context: &mut Context,
-        buf: &mut [u8],
+        mut buf: &mut [u8],
     ) -> Poll<io::Result<usize>> {
+        if self.fragmented_read {
```

Comment:
not sure if important but I think we're missing the edge-case where buf is of length 0 :P

Reply:
test code :P

Reply:
Added an explicit `assert!(!buf.is_empty());` to make this clearer. While it's not strictly an error to `read(&[])`, it's kind of undefined and isn't portable. So I've added this assert to help us avoid depending on this undefined behaviour.